### PR TITLE
2569 datatype generic

### PR DESCRIPTION
**Pull request by digitaldiina**
Created 1 new file in the datatypes section based on [Michel's blog post](https://community.superoffice.com/en/blog/introduction-to-crmscript-generic/)
Moved API reference files from global functions to a new Generic folder in classes.

### CODE REVIEW

File: Customer.Service.Support/CRMScript/datatypes/generic-type.md [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+* Byte GenericToByte(Generic generic)
+* Generic[] GenericToArray(Generic generic)
+
+Together with `getTypeName()`, these methods can be used to create an explicit typed variable.
+
+```crmscript
+Generic g = getCurrentDateTime();
+if (getTypeName(g)) {
+  DateTime dt = GenericToDateTime(g);
+}
```
Comment (AnthonyYates) [inline Customer.Service.Support/CRMScript/datatypes/generic-type.md:61]: This code does not run...
```suggestion
if (getTypeName(g) == "DateTime") {
  DateTime dt = GenericToDateTime(g);
}

```

File: Customer.Service.Support/CRMScript/datatypes/generic-type.md [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+
+## Generic and arrays
+
+### Integer getTypeDimensions(Generic generic)
+
+`getTypeDimensions()` returns the number of array dimensions for a variable. The argument is automatically up-casted to a Generic.
+
+```crmscript!
+Integer[][] i;
+printLine(getTypeDimensions(i));
```
Comment (AnthonyYates) [inline Customer.Service.Support/CRMScript/datatypes/generic-type.md:72]: This line of code does not run.
```suggestion
printLine(getTypeDimensions(i).toString());

```

File: Customer.Service.Support/CRMScript/datatypes/generic-type.md [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+```crmscript!
+struct Person {
+  String name;
+  Integer age;
+};
+
+Person p;
+
+String[] members = getStructMembers(p);
```
Comment (AnthonyYates) [inline Customer.Service.Support/CRMScript/datatypes/generic-type.md:122]: This code does not run.
```suggestion
String[] members = getStructMembers(getTypeName(p));

```

File: Customer.Service.Support/CRMScript/datatypes/generic-type.md [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+```
+
+### Generic getGenericValue(Generic struct, String name)
+
+A variant of `getGenericValue()` that returns a variable from the run-time environment **inside a struct** given its name and independent of its type.
+
+```crmscript
+Person p;
+p.age = 42;
+Generic g = getGenericValue(person, "age");
```
Comment (AnthonyYates) [inline Customer.Service.Support/CRMScript/datatypes/generic-type.md:136]:
```suggestion
Generic g = getGenericValue(p, "age");

```

File: Customer.Service.Support/CRMScript/datatypes/generic-type.md [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+
+### String convertGenericToString(Generic generic)
+
+`convertGenericToString()` returns the string representation of a variable.  The argument is automatically be up-casted to a Generic.
+
+This works for all **basic** types. Complex types might can't be serialized as easy, and might not support this conversion. In that case, they'll return “[complex]”.
+
+```crmscript!
+Integer age = 42;
+Generic g = convertGenericToString(age);
```
Comment (AnthonyYates) [inline Customer.Service.Support/CRMScript/datatypes/generic-type.md:150]:
```suggestion
String g = convertGenericToString(age);

```

**Review by AnthonyYates — CHANGES_REQUESTED** [review]
Please take a look at my comments and modify or accept the code example suggestions.

File: Customer.Service.Support/CRMScript/datatypes/generic-type.md (hunk 5 above)
Comment (digitaldiina) [reply Customer.Service.Support/CRMScript/datatypes/generic-type.md:150]: String s

File: Customer.Service.Support/CRMScript/datatypes/generic-type.md (hunk 3 above)
Comment (digitaldiina) [reply Customer.Service.Support/CRMScript/datatypes/generic-type.md:122]: getStructMembers takes the name of the struct, so p should be correct here

File: Customer.Service.Support/CRMScript/datatypes/generic-type.md [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+```
+
+### Generic getGenericValue(String name)
+
+`getGenericValue()` returns a variable from the run-time environment given its name and independent of its type.
+
+```crmscript!
+Integer age = 42;
+Generic g = getGenericValue("age");
+printLine(g);
```
[the remark this answers is not in the document]
Comment (digitaldiina) [reply Customer.Service.Support/CRMScript/datatypes/generic-type.md:34]: missing semicolon at the end of the revised line

File: Customer.Service.Support/CRMScript/datatypes/generic-type.md (hunk 5 above)
Comment (AnthonyYates) [reply Customer.Service.Support/CRMScript/datatypes/generic-type.md]: This does not seem to be changed. Please fix this and request a new review.

File: Customer.Service.Support/CRMScript/datatypes/generic-type.md (hunk 3 above)
Comment (AnthonyYates) [reply Customer.Service.Support/CRMScript/datatypes/generic-type.md]: This does not seem to be fixed. Please accept my suggestion, or change to:
String[] members = getStructMembers("Person");

**Review by AnthonyYates — CHANGES_REQUESTED** [review]
Still needs a couple code fixes.

**Review by AnthonyYates — APPROVED** [review]
Good to go.

**Pull request merged**